parquet-tools/src/main/java/org/apache/parquet/tools/command/DumpCommand.java

```diff
@@ -165,48 +165,47 @@ public static void dump(PrettyPrintWriter out, ParquetMetadata meta, MessageType
-                    out.rule('-');
-
-
-                    List<ColumnChunkMetaData> ccmds = block.getColumns();
-                    if (showColumns != null) {
-                        ccmds = new ArrayList<ColumnChunkMetaData>();
-                        for (ColumnChunkMetaData ccmd : block.getColumns()) {
-                            String path = Joiner.on('.').skipNulls().join(ccmd.getPath().toArray());
-                            if (showColumns.contains(path)) {
-                                ccmds.add(ccmd);
-                            }
-                        }
-                    }
-
-                    MetadataUtils.showDetails(out, ccmds);
+              if (group != 0)
+                out.println();
+              out.format("row group %d%n", group++);
+              out.rule('-');
+
+              List<ColumnChunkMetaData> ccmds = block.getColumns();
+              if (showColumns != null) {
+                ccmds = new ArrayList<ColumnChunkMetaData>();
+                for (ColumnChunkMetaData ccmd : block.getColumns()) {
+                  String path = Joiner.on('.').skipNulls().join(ccmd.getPath().toArray());
+                  if (showColumns.contains(path)) {
+                    ccmds.add(ccmd);
+                  }
+                }
+              }
 
-                    List<BlockMetaData> rblocks = Collections.singletonList(block);
-                    freader = new ParquetFileReader(
-                        conf, meta.getFileMetaData(), inpath, rblocks, columns);
-                    PageReadStore store = freader.readNextRowGroup();
-                    while (store != null) {
-                        out.incrementTabLevel();
-                        for (ColumnDescriptor column : columns) {
-                            out.println();
-                            dump(out, store, column);
-                        }
-                        out.decrementTabLevel();
+              MetadataUtils.showDetails(out, ccmds);
 
-                        store = freader.readNextRowGroup();
-                    }
-                    out.flushColumns();
+              List<BlockMetaData> rblocks = Collections.singletonList(block);
+              freader = new ParquetFileReader(conf, meta.getFileMetaData(), inpath, rblocks, columns);
+              PageReadStore store = freader.readNextRowGroup();
+              while (store != null) {
+                out.incrementTabLevel();
+                for (ColumnDescriptor column : columns) {
+                  out.println();
+                  dump(out, store, column);
                 }
+                out.decrementTabLevel();
+
+                store = freader.readNextRowGroup();
+              }
+              out.flushColumns();
             } finally {
-                if (freader != null) {
-                    freader.close();
-                }
+              if (freader != null) {
+                freader.close();
```

Comment:
If an I/O exception occurs in the `ParquetFileReader` constructor then you might try to close the previous reader object. I would guess there is no problem with invoking `close` multiple times on the same object but who knows? It might be a good idea to reset the reference so you would not close the same object again. Another idea would be to use try-with-resources for the code part where the reader is really needed.

Reply:
This is the memory used before and after fix this bug:
![image](https://user-images.githubusercontent.com/26733114/102300060-20513c00-3f8f-11eb-8d64-06844560419b.png)

Reply:
I did not say your fix is not working. What I suggest is to avoid closing a reader twice if an issue happens in the `ParquetFileReader` constructor.

Reply:
yeah yeah,  thx for your review, i am sorry, please forgive me. Do you mean we can fix it like this:
`
...
            } finally {
              if (freader != null) {
                freader.close();
                // reset current reader is null, avoid closing a reader twice
                freader = null;
              }
            }
`
or
`
try(ParquetFileReader freader = new ParquetFileReader(conf, meta.getFileMetaData(), inpath, rblocks, columns)){
// do something where the freader is really needed.
}
`
is right ?

Reply:
hi, @gszadovszky Could you help me to review this PR ? thanks.

Reply:
The `ParquetFileReader` is Closeable, I would suggest using the try-with-resource pattern: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

Reply:
@Fokko thanks for your suggestion, can you review it again?